src/main/java/org/jboss/threads/JBossThread.java

```diff
@@ -153,18 +153,27 @@ public void interrupt() {
         if (isInterrupted()) return;
         final AtomicInteger stateRef = this.stateRef;
         int oldVal, newVal;
-        do {
+        for (;;) {
             oldVal = stateRef.get();
-            if (oldVal == STATE_INTERRUPT_PENDING || oldVal == STATE_INTERRUPT_IN_PROGRESS) {
+            if (oldVal == STATE_INTERRUPT_PENDING) {
                 // already set
                 Messages.msg.tracef("Interrupting thread \"%s\" (already interrupted)", this);
                 return;
-            } else if (oldVal == STATE_INTERRUPT_DEFERRED) {
-                newVal = STATE_INTERRUPT_PENDING;
+            } else if (oldVal == STATE_INTERRUPT_IN_PROGRESS) {
+                // wait for interruption on other thread to be completed
+                JDKSpecific.onSpinWait();
+                continue;
```

Comment:
@dmlloyd What about JDK 8 with only one core? Could it be an endless loop?

Reply:
Not endless, because the thread could be rescheduled at some point.  But this is the most naive possible spin loop implementation.  A more robust loop would spin *n* times before falling back to `Thread.yield()`, where *n* is a constant number, but is 0 on single-core systems.  Since the fix seems to work, I'll update it accordingly.